### PR TITLE
Update documentation Build File to Kotlin

### PR DESCRIPTION
**Pull request by lkasso**

### CODE REVIEW

**Comment by lkasso** [comment]
@bot-gradle help

**Comment by lkasso** [comment]
@bot-gradle testBD

**Comment by lkasso** [comment]
@bot-gradle test BD

**Comment by lkasso** [comment]
@alllex, this old fix is failing:
https://github.com/gradle/gradle/blob/49f05dab360fa89c82dc1391d36029746d64182a/platforms/documentation/docs-asciidoctor-extensions-base/src/main/java/org/gradle/docs/asciidoctor/HeaderInjectingPostprocessor.java#L48

**Review by eskatos — CHANGES_REQUESTED** [review]
Good stuff 💪

File: platforms/documentation/docs/build.gradle.kts [hunk 1]
```diff
@@ -1,98 +1,115 @@
...
-    modules.pdf.version '2.3.10'
+    //    version = '2.5.13'
+    //    modules.pdf.version '2.3.10'
+    setVersion("2.5.13")
+    modules {
+        pdf.setVersion("2.3.10")
+    }
+    getModules().getPdf().version("2.3.10")
+    getModules().getPdf().use()
+    modules.pdf.version("2.3.10")
```
[the remark this answers is not in the document]
Comment (lkasso) [reply platforms/documentation/docs/build.gradle.kts:95]: Leftover from an experiment. Cleaned it up.

File: platforms/documentation/docs/build.gradle.kts [hunk 2]
```diff
@@ -113,472 +130,471 @@ tasks.named("stageDocs") {
...
+        val id = org.gradle.docs.internal.StringUtils.toLowerCamelCase("snippet-$categoryName-$snippetName")
         publishedSamples.create(id) {
             description = "Snippet from $snippetDir"
             category = "Other"
-            readmeFile = file("src/snippets/default-readme.adoc")
-            sampleDirectory = snippetDir
-            promoted = false
+            readmeFile.set(file("src/snippets/default-readme.adoc"))
+            sampleDirectory.set(snippetDir)
+            promoted.set(false)
```
[the remark this answers is not in the document]
Comment (lkasso) [reply platforms/documentation/docs/build.gradle.kts:244]: Updated.

File: platforms/documentation/docs/build.gradle.kts [hunk 3]
```diff
@@ -600,120 +616,117 @@ tasks.named("quickTest") {
 }
 
 // TODO add some kind of test precondition support in sample test conf
-tasks.named("docsTest") { task ->
+tasks.withType<Test>().named("docsTest") {
```
[the remark this answers is not in the document]
Comment (lkasso) [reply platforms/documentation/docs/build.gradle.kts:619]: Not sure what happened there. Fixed.

File: platforms/documentation/docs/build.gradle.kts [hunk 4]
```diff
@@ -1,98 +1,110 @@
...
 }
 
 asciidoctorj {
-    version = '2.5.13'
-    modules.pdf.version '2.3.10'
+    setVersion("2.5.13")
+    modules {
+        pdf.use()
+        pdf.setVersion("2.3.10")
+    }
```
Comment (alllex) [inline platforms/documentation/docs/build.gradle.kts:91]: ❌  Why do we call `.use()` ? What it does is sets the version to some default, but then we immediately override it.
We should be fine with just setting the version directly

```
    setVersion("2.5.13")
    modules.pdf.setVersion("2.3.10")
```

File: platforms/documentation/docs/build.gradle.kts [hunk 5]
```diff
@@ -600,120 +611,117 @@ tasks.named("quickTest") {
...
-            excludeTestsMatching "*WithoutCC*.sample"
-
-            // Projects generated by gradle init enable configuration cache by default, no need to run them again.
-            excludeTestsMatching "org.gradle.docs.samples.*building-*.sample"
+            excludeTestsMatching("org.gradle.docs.samples.*.snippet-configuration-cache-*.sample")
+            excludeTestsMatching("*WithoutCC*.sample")
 
             // These tests cover features that are not planned to be supported in the first stable release of the configuration cache.
-            def testsForUnsupportedFeatures = [
+            val testsForUnsupportedFeatures = arrayOf(
```
Comment (alllex) [inline platforms/documentation/docs/build.gradle.kts:724]: 💅  It's more canonical to use `listOf` in cases like this in Kotlin. (there is a couple of these below too)

File: platforms/documentation/docs/build.gradle.kts [hunk 6]
```diff
@@ -600,120 +611,117 @@ tasks.named("quickTest") {
...
         systemProperty("org.gradle.integtest.samples.cleanConfigurationCacheOutput", "true")
         systemProperty("org.gradle.integtest.executer", "configCache")
 
         filter {
             // Configuration cache samples enable configuration cache explicitly. We're not going to run them with the configuration cache executer.
-            excludeTestsMatching "org.gradle.docs.samples.*.snippet-configuration-cache-*.sample"
-            excludeTestsMatching "*WithoutCC*.sample"
-
-            // Projects generated by gradle init enable configuration cache by default, no need to run them again.
-            excludeTestsMatching "org.gradle.docs.samples.*building-*.sample"
```
Comment (alllex) [inline platforms/documentation/docs/build.gradle.kts:713]: ❓ We seem to have lost this exclude. Is this intentional?

**Review by alllex — CHANGES_REQUESTED** [review]

File: platforms/documentation/docs/build.gradle.kts (hunk 6 above)
Comment (lkasso) [reply platforms/documentation/docs/build.gradle.kts:713]: This must have been new since I started this PR. I made sure its in the Kotlin version.

**Comment by lkasso** [comment]
@bot-gradle test BD

**Comment by lkasso** [comment]
@bot-gradle clean

**Comment by lkasso** [comment]
@bot-gradle test BD

**Comment by lkasso** [comment]
@bot-gradle clean

**Comment by lkasso** [comment]
@bot-gradle test BD

File: platforms/documentation/docs/build.gradle.kts [hunk 7]
```diff
@@ -85,24 +85,47 @@ dependencies {
...
+
+tasks.named<AsciidoctorTask>("userguideSinglePageHtml").configure  {
+    val task = this
+    task.extensions.getByType<org.asciidoctor.gradle.jvm.AsciidoctorJExtension>().apply {
+        docExtensions(
+            project.dependencies.create(project(":docs-asciidoctor-extensions")),
+            project.dependencies.create(files("src/main/resources"))
+        )
+    }
+}
```
Comment (alllex) [inline platforms/documentation/docs/build.gradle.kts:113]: 🤔  Now there is quite a bit of duplication. Why was the solution with an if inside configureEach not appropriate?
I'd vote for removing the duplication. We could either have the original solution, or at least extract a common function that configured the extension and call it for the appropriate tasks. I'll leave it up to you to decide, I don't have a strong preference.

File: platforms/documentation/docs/build.gradle.kts [hunk 8]
```diff
@@ -85,24 +85,47 @@ dependencies {
...
             project.dependencies.create(project(":docs-asciidoctor-extensions-base"))
         )
-    } else {
-        asciidoctorj.docExtensions(
+    }
+}
+
+tasks.named<AsciidoctorTask>("userguideSinglePageHtml").configure  {
+    val task = this
+    task.extensions.getByType<org.asciidoctor.gradle.jvm.AsciidoctorJExtension>().apply {
```
Comment (alllex) [inline platforms/documentation/docs/build.gradle.kts:107]: For something more standard like `extensions` we probably don't need to qualify it with a `task.`

File: platforms/documentation/docs/build.gradle.kts [hunk 9]
```diff
@@ -85,24 +85,47 @@ dependencies {
...
         Regex("^(?!GIF image format not supported|dropping cells from incomplete row detected end of table|.*Asciidoctor PDF does not support table cell content that exceeds the height of a single page).*").toPattern()
     )
 }
 
-tasks.withType<AsciidoctorTask>().configureEach {
-    if (name == "userguideSinglePagePdf") {
-        asciidoctorj.docExtensions(
+tasks.named<AsciidoctorTask>("userguideSinglePagePdf").configure {
+    val task = this
+    task.extensions.getByType<org.asciidoctor.gradle.jvm.AsciidoctorJExtension>().apply {
```
Comment (alllex) [inline platforms/documentation/docs/build.gradle.kts:98]: 💅  There is no need for `apply` as we can call `docExtensions` directly

**Review by alllex — APPROVED** [review]
I left some comments for polishing

File: platforms/documentation/docs/build.gradle.kts (hunk 7 above)
Comment (lkasso) [reply platforms/documentation/docs/build.gradle.kts:113]: I thought that's what you meant when you said:
>  could you also make sure the task is explicit?
I misunderstood.

Comment (lkasso) [reply platforms/documentation/docs/build.gradle.kts:113]: Reverting to the old style.

**Pull request merged**